android/sdk/src/main/java/org/jitsi/meet/sdk/connection_service/ConnectionService.java

```diff
@@ -0,0 +1,143 @@
+package org.jitsi.meet.sdk.connection_service;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Bundle;
+import android.support.annotation.RequiresApi;
+import android.telecom.Connection;
+import android.telecom.ConnectionRequest;
+import android.telecom.PhoneAccount;
+import android.telecom.PhoneAccountHandle;
+import android.telecom.TelecomManager;
+import android.util.Log;
+
+import com.facebook.react.bridge.Promise;
+
+/**
+ * Jitsi Meet implementation of {@link ConnectionService}. At the time of this writing it implements
+ * only the outgoing call scenario.
+ *
+ * @author Pawel Domas
+ */
+@RequiresApi(api = Build.VERSION_CODES.O)
+public class ConnectionService extends android.telecom.ConnectionService {
+
+    /**
+     * Tag used for logging.
+     */
+    public static final String TAG = "JitsiConnectionService";
+
+    /**
+     * The extra added to the {@link ConnectionImpl} and {@link ConnectionRequest} which stores
+     * the call's UUID.
+     */
+    static final String EXTRAS_CALL_UUID = "org.jitsi.meet.CALL_UUID";
+
+    /**
+     * The extra added to the {@link ConnectionImpl} and {@link ConnectionRequest} which stores
+     * the {@link PhoneAccountHandle} created for the call.
+     */
+    static final String EXTRA_PHONE_ACCOUNT_HANDLE
+        = "org.jitsi.meet.sdk.connection_service.PHONE_ACCOUNT_HANDLE";
+
+    @Override
+    public Connection onCreateOutgoingConnection(
+            PhoneAccountHandle accountHandle, ConnectionRequest request) {
+        ConnectionImpl connection = new ConnectionImpl(this);
+
+        connection.setConnectionProperties(Connection.PROPERTY_SELF_MANAGED);
+        connection.setAddress(request.getAddress(), TelecomManager.PRESENTATION_ALLOWED);
+        connection.setExtras(request.getExtras());
+
+        Bundle moreExtras = new Bundle();
+
+        moreExtras.putParcelable(EXTRA_PHONE_ACCOUNT_HANDLE, request.getAccountHandle());
+        connection.putExtras(moreExtras);
+
+        ConnectionList.getInstance().add(connection);
+        Promise startCallPromise
+            = ConnectionList.getInstance().unregisterStartCallPromise(connection.getCallUUID());
+
+        if (startCallPromise != null) {
+            startCallPromise.resolve(null);
+        } else {
+            Log.e(TAG,
+                String.format(
+                        "onCreateOutgoingConnection: no start call Promise for %s",
+                        connection.getCallUUID()));
+        }
+
+        // TODO Implement video state. Currently there's not much use of it without the incoming
```

Comment:
How come? Maybe something in the connected service behaves differently depending on the call type? Even on outgoing calls, we do know if video is used or not.

Reply:
As far as my understanding of a self managed ConnectionService goes it will not have any effect, but I can still do that anyway if you insist.

Reply:
So as per our discussion offline it's probably needed, because the system may decide to use the speaker with video

Reply:
So my device doesn't automatically select the speaker for video calls, but maybe I'm missing something. It could also depend on the device. Anyway it's good to have it I guess.